### ISSUE TITLE
Turning off auto-fusing breaks Http.singleRequest

### ISSUE THREAD

**Issue by ktoso** [akka/akka-http#548]
<a href="https://github.com/lilac"><img src="https://avatars.githubusercontent.com/u/705652?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [lilac](https://github.com/lilac)**
_Friday Nov 18, 2016 at 01:50 GMT_
_Originally opened as https://github.com/akka/akka/issues/21861_

----

While debugging an issue of `Http.singleRequest` hang (timeout), I finally found that is caused by `auto-fusing` turned off. This is quite a surprise. Below is the snippet that reproduces this unexpected behaviour (or bug).

```scala
package com.samemoment

import scala.concurrent.Future

import akka.actor.ActorSystem
import akka.http.scaladsl.Http
import akka.http.scaladsl.model.{ HttpRequest, HttpResponse }
import akka.stream.ActorMaterializer
import akka.util.Timeout
import com.typesafe.config.ConfigFactory

/**
 * Copyright SameMo 2016
 */
object FuseTest {
  val config = ConfigFactory.parseString(
    s"""
       |akka {
       |  stream {
       |    materializer {
       |      auto-fusing: off
       |    }
       |  }
       |}
    """.stripMargin)

  implicit val system = ActorSystem("test", config)
  implicit val materializer = ActorMaterializer()
//  private implicit val timeout = Timeout(20.seconds)

  implicit val ec = system.dispatcher
  implicit val http = Http()

  def main(args: Array[String]): Unit = {
    val responseFuture: Future[HttpResponse] =
      Http().singleRequest(HttpRequest(uri = "https://api.twilio.com"))

    responseFuture map {
      r:
      HttpResponse =>
        println(r)
    }
    responseFuture andThen { case _ =>
      http.shutdownAllConnectionPools()
      system.terminate()
    }
  }
}
``` 

When auto-fusing is switched on, the request succeeds and the response got printed. But when it is off, the future never got complete, causing it hangs forever or timeout.

**Comment by ktoso** [comment]
<a href="https://github.com/drewhk"><img src="https://avatars.githubusercontent.com/u/932285?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [drewhk](https://github.com/drewhk)**
_Friday Nov 18, 2016 at 08:02 GMT_

----

Hi @lilac, this looks like a weird issue indeed. The ticket needs to be moved though as Akka HTTP is now a separate project: https://github.com/akka/akka-http

**Comment by ktoso** [comment]
Though perhaps it should be considered if the client should take care of it's own materializer.

**Comment by michkhol** [comment]
With Scala 2.12.0, akka-actor 2.4.14 and akka-http-core 10.0.0 auto-fusing has no effect: it hangs all the time.

**Comment by michkhol** [comment]
In the configuration `Stream -> ActorSubscriber -> Router -> n Workers` it starts to hang where `n` is sufficiently large, like greater than 30. Tried it both ways, HttpExt instance shared by all workers and one HttpExt instance per worker. `auto-fusing` has no effect.

**Comment by ktoso** [comment]
Why are you using ActorSubscriber (no-one should use it really...)?
What's the exact layout you're doing?
ActorSubscriber is not impacted by fusing of course since it by definition wants to be an Actor.

**Comment by michkhol** [comment]
I'm using the Sink.actorSubscriber() to back-pressure a Stream with millions of XML fragments. The fragments are loaded to a service using REST API. The ActorSubscriber descendant creates a SmallestMailboxPool router with `n` workers. Each worker makes several http requests to the service per fragment. The requests are forced to be synchronous, i.e. I wait for all HttpResponse futures to complete inside the message handler (there are reasons for this outside the scope of this issue). 
After upgrading to akka 2.4.14 and akka-http 10.0.0 it started to hang on a sufficient load. If I compile with Scala 2.12.0, it hangs immediately.

**Comment by ktoso** [comment]
I strongly believe there's far too much your own code, and protocol to blame it purely on the scala version.
Please prepare an independent *minimised* reproducer if you seek help here, preferably as separate ticket - thanks!

**Comment by michkhol** [comment]
Working on it.

**Comment by michkhol** [comment]
Successfully reproduced it with the code minimized to a bare bones thread pool: https://github.com/akka/akka-http/issues/602

**Comment by jrudolph** [comment]
Might have been related to https://github.com/akka/akka/issues/22585.

**Comment by jrudolph** [comment]
Is this still an issue after the fix in akka/akka#22585? I'm going to close it for now, if that is still an issue with a version >= 10.0.9 and Akka >= 2.4.19 or >= 2.5.3, please let us know here or create a new ticket.
